### PR TITLE
Randomise the nonce in the block header

### LINKED ISSUE #1033: Multi-threaded mining via command-line option -genproclimit duplicates work by using the same nonce values
The zcash-cli option -genproclimit spawns multiple mining threads, but each one will use the same sequence of nonce and extraNonce values, thus duplicating work.

Currently, each thread starts with its own extraNonce at 0 (on the first run it is increased to 1) and this value is incremented on each new attempt to find a solution after a nonce has been exhausted.

Ideally there should be a generator for nonce and extranonce values (for any block being mined on a given prevblockhash) which all threads consume from to ensure work is unique.

To verify this, modify logging statement:

```
diff --git a/src/miner.cpp b/src/miner.cpp
index 2f9caf0..8d83a32 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -512,8 +512,8 @@ void static BitcoinMiner(CWallet *pwallet)
                                                       pblock->nNonce.size());
 
                     // (x_1, x_2, ...) = A(I, V, n, k)
-                    LogPrint("pow", "Running Equihash solver with nNonce = %s\n",
-                             pblock->nNonce.ToString());
+                    LogPrint("pow", "Running Equihash solver with nExtraNonce = %d and nNonce = %s\n",
+                             nExtraNonce, pblock->nNonce.ToString());
                     std::set<std::vector<unsigned int>> solns;
                     EhOptimisedSolve(n, k, curr_state, solns);
                     LogPrint("pow", "Solutions: %d\n", solns.size());

```

Launch zcashd:

`./src/zcashd -daemon -genproclimit=3 -debug=pow
`

And examine debug.log with:

`grep nExtraNonce debug.log
`

Output will look like this:
```
2016-06-18 22:25:41 Running Equihash solver with nExtraNonce = 1 and nNonce = 0000000000000000000000000000000000000000000000000000000000000000
2016-06-18 22:25:41 Running Equihash solver with nExtraNonce = 1 and nNonce = 0000000000000000000000000000000000000000000000000000000000000000
2016-06-18 22:25:41 Running Equihash solver with nExtraNonce = 1 and nNonce = 0000000000000000000000000000000000000000000000000000000000000000
```

### PR DESCRIPTION
**Pull request by str4d**
The top and bottom 16 bits of the nonce are left clear for local use as thread
flags and counters. This does not leak any more local information about the
miner than is currently exposed.

The cleared bits should not be considered a consensus rule, as miners are free
to set all bits of the nonce however they wish.

Closes #1033

### CODE REVIEW

File: src/miner.cpp [hunk 1]
```diff
@@ -352,12 +353,18 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
         pblock->vtx[0] = txNew;
         pblocktemplate->vTxFees[0] = -nFees;
 
+        // Randomise nonce
+        arith_uint256 nonce = UintToArith256(GetRandHash());
```
Comment (str4d) [inline src/miner.cpp]: Are there any security concerns with using `GetRandHash()` directly?

Comment (ebfull) [inline src/miner.cpp]: I'm not aware of any, it does use OpenSSL's random number generator though.

**Comment by ebfull** [comment]
I tested this locally. Works!

**Comment by daira** [comment]
Reviewing now.

**Comment by daira** [comment]
utACK, but the commit message should mention that it now updates the chain every solution run, rather than every second run.

**Comment by ebfull** [comment]
@zkbot r+

**Pull request merged**